### PR TITLE
RGB2RGB conversions rewritten to wide intirinsics

### PR DESCRIPTION
**Pull request by savuor**
### This pullrequest changes

Channel swap and 3/4 channel conversions rewritten for all data types using universal wide intrintsics.

### Performance

<details>

<summary>Baseline AVX2</summary>

|Name of Test| initial | wide | wide vs initial (x-factor)|
|---|:-:|:-:|:-:|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2BGRA)|0.008|0.003|2.56|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGB)|0.007|0.003|2.25|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGBA)|0.008|0.003|2.64|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2BGR)|0.007|0.003|2.64|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2RGBA)|0.008|0.003|2.51|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_RGBA2BGR)|0.007|0.003|2.57|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2BGRA)|0.318|0.080|3.99|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGB)|0.275|0.078|3.55|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGBA)|0.318|0.081|3.91|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2BGR)|0.275|0.081|3.39|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2RGBA)|0.334|0.096|3.48|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_RGBA2BGR)|0.275|0.083|3.29|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2BGRA)|2.164|0.687|3.15|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGB)|1.888|0.588|3.21|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGBA)|2.292|0.694|3.30|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2BGR)|2.054|0.721|2.85|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2RGBA)|2.358|0.873|2.70|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_RGBA2BGR)|1.990|0.710|2.80|

</details>

<details>
<summary>Baseline AVX512SKX</summary>

|Name of Test| initial | wide | wide vs initial (x-factor)|
|---|:-:|:-:|:-:|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2BGRA)|0.009|0.003|2.69|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGB)|0.008|0.003|2.38|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGBA)|0.009|0.004|2.44|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2BGR)|0.008|0.003|2.59|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2RGBA)|0.009|0.004|2.49|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_RGBA2BGR)|0.008|0.003|2.60|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2BGRA)|0.328|0.084|3.90|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGB)|0.313|0.082|3.84|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGBA)|0.327|0.083|3.93|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2BGR)|0.314|0.087|3.60|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2RGBA)|0.375|0.098|3.82|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_RGBA2BGR)|0.315|0.083|3.79|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2BGRA)|2.261|0.725|3.12|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGB)|1.933|0.611|3.16|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGBA)|2.237|0.726|3.08|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2BGR)|2.014|0.752|2.68|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2RGBA)|2.446|0.881|2.78|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_RGBA2BGR)|1.985|0.750|2.65|

</details>

<details>
<summary>Baseline SSE2</summary>

|Name of Test| initial | wide | wide  vs initial (x-factor)|
|---|:-:|:-:|:-:|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2BGRA)|0.008|0.004|2.13|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGB)|0.007|0.006|1.14|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGBA)|0.008|0.004|2.08|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2BGR)|0.007|0.005|1.25|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2RGBA)|0.008|0.003|2.52|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_RGBA2BGR)|0.007|0.005|1.30|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2BGRA)|0.319|0.142|2.24|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGB)|0.292|0.246|1.19|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGBA)|0.342|0.139|2.46|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2BGR)|0.300|0.228|1.31|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2RGBA)|0.349|0.128|2.72|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_RGBA2BGR)|0.298|0.228|1.30|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2BGRA)|2.180|1.060|2.06|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGB)|1.888|1.757|1.08|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGBA)|2.176|1.082|2.01|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2BGR)|1.948|1.575|1.24|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2RGBA)|2.233|0.989|2.26|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_RGBA2BGR)|1.940|1.524|1.27|

</details>

<details>
<summary>Baseline SSE4.1</summary>

|Name of Test| initial | wide | wide  vs  initial (x-factor)|
|---|:-:|:-:|:-:|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2BGRA)|0.008|0.002|3.69|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGB)|0.007|0.002|3.15|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGR2RGBA)|0.008|0.002|3.69|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2BGR)|0.007|0.003|2.82|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_BGRA2RGBA)|0.008|0.003|2.59|
|cvtColor8u::Size_CvtMode::(127x61, COLOR_RGBA2BGR)|0.007|0.003|2.66|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2BGRA)|0.327|0.079|4.14|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGB)|0.288|0.080|3.60|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGR2RGBA)|0.327|0.081|4.05|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2BGR)|0.289|0.094|3.08|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_BGRA2RGBA)|0.334|0.119|2.80|
|cvtColor8u::Size_CvtMode::(640x480, COLOR_RGBA2BGR)|0.289|0.093|3.10|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2BGRA)|2.238|0.694|3.23|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGB)|1.964|0.587|3.34|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGR2RGBA)|2.260|0.699|3.23|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2BGR)|1.990|0.788|2.53|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_BGRA2RGBA)|2.317|0.991|2.34|
|cvtColor8u::Size_CvtMode::(1920x1080, COLOR_RGBA2BGR)|2.022|0.792|2.55|

</details>

```
#disable_ipp=ON
```

### CODE REVIEW

**Comment by alalek** [comment]
1. Why master?
2. Please re-measure performance after tests are get fixed.
3. Use standard OpenCV tools for reports generation: summary.py

**Comment by savuor** [comment]
@alalek 
1. Sure, 3.4 is better. Branch changed.
2. Done.
3. I used these tools, the table is just used for convenience. The report from summary.py added to the PR description.

File: modules/imgproc/src/color_rgb.cpp [hunk 1]
```diff
@@ -12,183 +12,118 @@ namespace cv
...
+        n *= srccn;
+#if CV_SIMD
+        const int vsize = vt::nlanes;
+        for(; i <= n-vsize*srccn;
+            i += vsize*srccn, src += vsize*srccn, dst += vsize*dstcn)
         {
-            n *= 3;
-            if (scn == 3)
+            vt a, b, c, d;
+            if(srccn == 4)
```
[the remark this answers is not in the document]
Comment (savuor) [reply modules/imgproc/src/color_rgb.cpp]: discussed, decided to leave the codes as is (due to little performance benefit in comparison to code bloating)

**Comment by savuor** [comment]
@alalek 
1. The problem seems to be in `vx_cleanup()` code which wraps `_mm256_zeroupper` intrinsic. This breaks the tests even if there's no other intrinsics in the code. Threw it away, hope this helps.
2. Done
3. Done

**Pull request merged**